catalogue_pipeline/matcher/src/main/scala/uk/ac/wellcome/platform/matcher/locking/DynamoRowLockDao.scala

```diff
@@ -37,46 +40,59 @@ class DynamoRowLockDao @Inject()(
   def lockRow(id: Identifier, contextId: String) = Future {
     val (created, expires) = getExpiry
     val rowLock = RowLock(id.id, contextId, created, expires)
-
-    debug(s"Trying to create RowLock: $rowLock")
+    debug(s"Locking $rowLock")
 
     val scanamoOps = table
       .given(
         not(attributeExists('id)) or Condition(
           'expires < rowLock.created.getEpochSecond))
       .put(rowLock)
-
     val result = Scanamo.exec(dynamoDBClient)(scanamoOps)
-
-    debug(s"Got $result when creating $rowLock")
+    debug(s"Got $result for $rowLock")
 
     result match {
       case Right(_) => rowLock
-      case Left(error) =>
-        throw new FailedLockException(s"Failed to lock $id", error)
+      case Left(error) => {
+        debug(s"Failed to lock $rowLock $error")
+        throw FailedLockException(s"Failed to lock $id", error)
+      }
     }
   }
 
-  def unlockRow(contextId: String) = Future {
+  def unlockRows(contextId: String) = Future {
     debug(s"Trying to unlock context: $contextId")
-
-    val maybeRowLocks =
-      Scanamo.queryIndex[RowLock](dynamoDBClient)(table.name, index)(
-        'contextId -> contextId)
-    val rowLockIds = maybeRowLocks.collect {
-      case Right(rowLock) => rowLock.id
-    }.toSet
-
-    maybeRowLocks.collect {
-      case Left(error) => {
-        info(s"Error $error when unlocking $contextId")
-        throw FailedUnlockException(s"Failed to unlock [$contextId] $error")
+    try {
+      val maybeRowLocks: immutable.Seq[Either[DynamoReadError, RowLock]] =
+        Scanamo.queryIndex[RowLock](dynamoDBClient)(table.name, index)(
+          'contextId -> contextId)
+
+      debug("maybeRowLocks: " + maybeRowLocks)
+      val rowLockIds = maybeRowLocks.collect {
+        case Right(rowLock) => rowLock.id
+        case Left(error) =>
+          info(s"Error $error when unlocking $contextId")
+          throw FailedLockException(s"Failed to unlock [$contextId] $error")
+      }.toSet
+
+      debug(s"Unlocking rows: $rowLockIds")
+      val deleteAllResults =
+        Scanamo.deleteAll(dynamoDBClient)(table.name)('id -> rowLockIds)
+      deleteAllResults.foreach { result: BatchWriteItemResult =>
+        if (result.getUnprocessedItems.size() > 0) {
+          val error =
+            s"Batch delete failed to delete ${result.getUnprocessedItems}"
+          info(error)
+          throw FailedLockException(error)
```

Comment:
Would it be useful to reintroduce `FailedUnlockException`? If we want to add metrics it might be useful to distinguish them. A failure locking is fine, but a failure unlocking is something that is a bit harder to recover from, so we might want to know how often it happens